pkg/controllers/nodeclass/validation.go

```diff
@@ -189,18 +189,26 @@ func (v *Validation) validateRunInstancesAuthorization(
 	nodeClaim *karpv1.NodeClaim,
 	tags map[string]string,
 ) (reason string, err error) {
-	// NOTE: Since we've already validated the AMI status condition is true, this should never occur
+	// NOTE: Since we've already validated the status conditions are true, these should never occur
 	if len(nodeClass.Status.AMIs) == 0 {
 		return "", fmt.Errorf("no resolved amis in status")
 	}
+	if len(nodeClass.Status.Subnets) == 0 {
+		return "", fmt.Errorf("no resolved subnets in status")
+	}
+	if len(nodeClass.Status.SecurityGroups) == 0 {
+		return "", fmt.Errorf("no resolved security groups in status")
+	}
+	if nodeClass.Status.InstanceProfile == "" {
+		return "", fmt.Errorf("no instance profile in status")
+	}
 
 	var instanceType ec2types.InstanceType
 	requirements := scheduling.NewNodeSelectorRequirements(nodeClass.Status.AMIs[0].Requirements...)
 
+	instanceType = ec2types.InstanceTypeM6gLarge
```

Comment:
Is this hard-coded because we run this validation before Karpenter runs simulation and determines the instance type?

Not really a blocker for this fix but I wonder what's the downside of this: can we end up blocking customers if say they are at their service quota limit for this instance type (even though Karpenter isn't really going to spin this up)?

Reply:
> Maybe I am missing some context here but the RunInstances API considers this field as optional

If you don't specify an instance type and try to use an ARM AMI ID, it will fail because the arch doesn't match

Reply:
Sorry for the drive-by comment.. 🙄

What is the actual value of the dry-run call in the first place? Given that it's possible to not cover edge cases as discussed here and seen in #7834 why not simply fail in the actual CreateFleet call?

Does it bring better error messages compared to CreateFleet or what is the motivation?

Reply:
Trying it out on my side, it doesn't appear that EC2 does anything smart here -- they basically launch every time with `m1.small` if you don't specify any instance type, which is probably more rare than the m5 we statically coded here

Reply:
> it's possible to not cover edge cases as discussed here

I'm not sure that I agree that we can't cover the edge cases. The motivation here is to get a quick and clear error response (and indication on Karpenter objects) that something is wrong with the system